glyphs-reader/src/glyphdata.rs

```diff
@@ -3,145 +3,379 @@
+                ))
+            }
         }
     }
 
+    // now we've found some values, let's finalize them
+    let name = name
+        .map(SmolStr::new)
+        .ok_or_else(|| GlyphDataError::missing_attr("name", item.attributes_raw()))?;
+    let category = category
+        .ok_or_else(|| GlyphDataError::missing_attr("category", item.attributes_raw()))
+        .and_then(|cat| {
+            Category::from_str(cat.as_ref()).map_err(GlyphDataError::InvalidCategory)
+        })?;
+    let subcategory = subcategory
+        .map(|cat| Subcategory::from_str(cat.as_ref()).map_err(GlyphDataError::InvalidSubcategory))
+        .transpose()?;
+    let codepoint = unicode
+        .map(|s| {
+            u32::from_str_radix(&s, 16).map_err(|inner| GlyphDataError::InvalidUnicode {
+                raw: s.into_owned(),
+                inner,
+            })
+        })
+        .transpose()?;
+    let alt_names = alt_names
+        .map(|names| {
+            names
+                .as_ref()
+                .split(',')
+                .map(|name| SmolStr::from(name.trim()))
+                .collect()
+        })
+        .unwrap_or_default();
+
+    Ok(GlyphInfoFromXml {
+        name,
+        alt_names,
+        category,
+        subcategory,
+        codepoint,
+    })
+}
+
+impl GlyphData {
     /// Get the info for the given name/codepoints, attempting to synthesize it if necessary
     ///
-    /// If this name or these unicode values were included in the bundled data,
-    /// that will be returned; otherwise we will attempt to compute the value
-    /// by performing various heuristics based on the name.
+    /// Returns, from most to least preferred:
+    ///
+    /// 1. The matching override value
+    /// 1. The matching value from bundled data
+    /// 1. A computed value based on name heuristics
     ///
     // See https://github.com/googlefonts/glyphsLib/blob/e2ebf5b517d/Lib/glyphsLib/glyphdata.py#L94
-    pub fn get_glyph(
+    pub fn query(
         &self,
         name: &str,
         codepoints: Option<&BTreeSet<u32>>,
-    ) -> Option<Cow<GlyphInfo>> {
-        if let Some(info) = self.get_by_name(name).or_else(|| {
-            codepoints
-                .into_iter()
-                .flat_map(|cps| cps.iter())
-                .find_map(|cp| self.get_by_codepoint(*cp))
-        }) {
-            return Some(Cow::Borrowed(info));
-        }
-
-        // we don't have info for this glyph: can we synthesize it?
-        // TODO: python does production name here.
-        // see https://github.com/googlefonts/fontc/issues/780
-
-        let (category, subcategory) = self.construct_category(name)?;
-        Some(Cow::Owned(GlyphInfo {
-            name: name.into(),
-            category,
-            subcategory,
-            unicode: None,
-            production: None,
-            alt_names: Default::default(),
-        }))
+    ) -> Option<(Category, Option<Subcategory>, Option<u32>)> {
```

Comment:
A little struct or something might be useful here, I can't tell at a glance what the `u32` is supposed to be.

Reply:
Good point. It maybe made more sense when it was just `Option<(Category, Subcategory)>` :D

Reply:
Fixed